### ISSUE TITLE
Dockerfile

### ISSUE THREAD

**Issue by grahovam** [diaspora/diaspora#6846]
Why isn't there a Dockerfile yet? Has there been a discussion about that before? I could make one and add installation guides on how to Install Diaspora using Docker.

**Comment by jhass** [comment]
https://github.com/search?utf8=%E2%9C%93&q=diaspora+docker

**Comment by grahovam** [comment]
Yes, I have seen those too. But for many people, Docker is the first choice when it's about trying out something they don't know. And they go into the installation guides section and find out that there is no Docker guide. And no Dockerfile in this repo. So let this be the discussion about if we want a Dockerfile in this repo or not. 

I don't want to find out which one of the Dockerfiles out there is the best, before testing Diaspora on my VPS. I don't even know if I like Diaspora. I want to have a Dockerfile by the Diaspora community in this repo.

**Comment by jhass** [comment]
As far as I'm aware non of the core devs actually uses Docker, not for development and neither for deployment. Who should maintain it in the repo?

**Comment by denschub** [comment]
I have to agree with @jhass. Pulling Docker configurations into the official repo would need someone to constantly support it. We have some notes about Capistrano, Vagrant, Puppy and Ansible [in the wiki](https://wiki.diasporafoundation.org/Installation), so feel free to add any Docker configurations in an own Docker-based installation guide, but for now, Docker is nothing I feel like we can support.

**Comment by Flaburgan** [comment]
Nobody currently officially supporting doesn't mean nobody will nor nobody wants to. To add support of docker and to find a maintainer are two different issues. To close this issue means we don't want to support docker. If we agree to support it and are looking for a maintainer, then we should do the opposite: let this issue and clearly indicate that we're looking for someone.

About this specific problem, @Chocobozzz is a french diaspora* user available everyday on IRC, if we decide to support docker, we can easily ask him if he wants to be the maintainer.

**Comment by denschub** [comment]
> To close this issue means we don't want to support docker.

If you think that's the case, fine, I'll reopen.

**Comment by robbyoconnor** [comment]
Why not do something like [discourse did](https://github.com/discourse/discourse_docker)? 

Here is a [blog post by Sam Saffron](https://samsaffron.com/archive/2013/11/07/discourse-in-a-docker-container) about how he did it and why.

**Comment by jhass** [comment]
@robbyoconnor great, are you going to write and maintain it?

**Comment by robbyoconnor** [comment]
My time right now is uh...limited, just giving an idea if somebody else who is following this issue wants to take it on, they may :) I've been following the project quietly :smiley_cat:

**Comment by goobertron** [comment]
Brad Koehn is currently working on a Docker set-up for Diaspora: https://gitlab.koehn.com/bkoehn/diaspora-docker

So perhaps this will become an 'official thing' in a while.

**Comment by dhovart** [comment]
If I can be of any help, I use docker for my development environment.
I can update the wiki with what I've already written [here](https://github.com/dhovart/diaspora-docker) if that sounds good to you.

**Comment by robbyoconnor** [comment]
These should be officially documented. I should note however that Docker in development is good for end-to-end testing but typically I just run the services in docker and then run the app itself locally. It's not like vagrant in any way...on some systems it may have used virtualbox but on Linux it's not that great to run the app itself inside of Docker.

**Comment by dhovart** [comment]
Well, it's a good way to get started while keeping your system clean: you won't need to install all of the dependencies required by the project.
I'm not sure I see your point. If you can install the services with docker, then why wouldn't you want to do it for the app as well ?

**Comment by robbyoconnor** [comment]
I'm not sure -- I guess I just never figured it out...

**Comment by dhovart** [comment]
That's what volumes are for: you can mount a volume in your container that points to a host directory. I work on the code of diaspora in  host directly and the changes are reflected immediately in the container.
In the Dockerfile I create the [/diaspora](https://github.com/dhovart/diaspora-docker/blob/master/Dockerfile#L8) directory and set it to mount my [current directory](https://github.com/dhovart/diaspora-docker/blob/master/docker-compose.yml#L12) (where is the diaspora code) when starting the "web" container.
https://docs.docker.com/engine/tutorials/dockervolumes/#/mount-a-host-directory-as-a-data-volume

**Comment by dhovart** [comment]
Docker-compose makes all even easier, as it lets you specify how your containers running the required services need to interact together and which image to use (for instance there are official images for redis or postgresql). Nothing else to do : it pulls the specified images, expose the ports you tell it to, run a default command to start whatever service after the containers boot and there, your env is ready.

For instance if I'm tempted to check the chat features (I haven't yet) I think I could just add prosody to the mix without it being too painful, and without messing up my system.

**Comment by robbyoconnor** [comment]
Nice! I guess I never looked into it -- thanks! Learned something new @dhovart !!! By the way nice job!

**Comment by robbyoconnor** [comment]
Are you on Linux @dhovart ?

**Comment by dhovart** [comment]
Sorry for the late answer, I was travelling. I'm OS X. NFS sharing is very slow, so I use this little script instead: https://github.com/brikis98/docker-osx-dev

**Comment by robbyoconnor** [comment]
yeah -- I'm running natively on Linux without a VM in the middle.
